src/coreclr/jit/valuenum.cpp

```diff
@@ -7793,10 +7793,44 @@ ValueNum ValueNumStore::EvalHWIntrinsicFunBinary(var_types      type,
             }
 
 #ifdef TARGET_ARM64
-            case NI_AdvSimd_Multiply:
             case NI_AdvSimd_MultiplyByScalar:
-            case NI_AdvSimd_Arm64_Multiply:
             case NI_AdvSimd_Arm64_MultiplyByScalar:
+            {
+                if (!varTypeIsFloating(baseType))
+                {
+                    // Handle `x * 0 == 0` and `0 * x == 0`
+                    // Not safe for floating-point when x == -0.0, NaN, +Inf, -Inf
+                    ValueNum zeroVN = VNZeroForType(TypeOfVN(cnsVN));
+
+                    if (cnsVN == zeroVN)
+                    {
+                        return VNZeroForType(type);
+                    }
+                }
+
+                // Handle x * 1 => x, but only if the scalar RHS is 1.
+                ValueNum oneVN;
+                if (varTypeIsSIMD(TypeOfVN(arg1VN)))
+                {
+                    oneVN = VNOneForSimdType(TypeOfVN(arg1VN), baseType);
+                }
+                else
+                {
+                    oneVN = VNOneForType(baseType);
+                }
+
+                if (arg1VN == oneVN)
+                {
+                    assert(TypeOfVN(arg0VN) == type);
+                    return arg0VN;
+                }
```

Comment:
This looks correct, but may be slightly pessimizing for `MultiplyByScalar`.

While we try to self-import/normalize `vector * scalar` as `MultiplyByScalar(vector, Vector64.CreateScalarUnsafe(scalar))` and `CreateScalarUnsafe` will itself try to create a broadcast for a constant `scalar` (for efficiency reasons), a user could still call `MultiplyByScalar` themselves among other factors.

So I think the more correct thing to do here is to expect that `arg1VN` is a vector (check and exit if it isn't), then to extract element 0 and check if that is equal to `VNOneForType(baseType)`. Since `Vector64.Create(1, 0)` will also do the same thing here.

--------------

Alternatively, maybe the better thing would be to make `MultiplyByScalar` a lowering optimization. We could import it as `Multiply(vector, Vector128.Create(scalar))` and have lowering look for this pattern and swap it around. That would allow more CSE opportunities and less complex handling here.

Reply:
-- I don't have a particular preference and aren't even particularly concerned about the slight pessimization.

Just wanted to call it out as something to consider since this code is being touched.

Reply:
Does it look better now?